### PR TITLE
vcd-readme

### PR DESCRIPTION
**Pull request by victordibia**
Update readme, add links to jump to method sections.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # config file
 config.js
+samples
```
Comment (jweisz) [inline .gitignore]: why is this necessary?

File: README.md [hunk 2]
```diff
@@ -349,7 +349,7 @@ tj.see().then(function(objects) {
 Sample resposne:
```
Comment (jweisz) [inline README.md]: just saw a typo in 'resposne' (this is my fault).. would you mind fixing it?  :D

File: lib/tjbot.js [hunk 3]
```diff
@@ -806,6 +806,8 @@ TJBot.prototype.read = function() {
  * Returns the photo data in a Buffer.
  */
 TJBot.prototype._captureImage = function(filePath) {
+    this._assertCapability('see');
```
Comment (jweisz) [inline lib/tjbot.js]: by current convention, we don't need to assert capabilities in private helper methods. end users should not be calling methods prefixed with a `_`, so there's no need to assert the capability here.

File: .gitignore (hunk 1 above)
Comment (victordibia) [inline .gitignore]: had some files in samples used for testing. This ensures they dont get added

File: lib/tjbot.js (hunk 3 above)
Comment (victordibia) [inline lib/tjbot.js]: This has been moved to the takePhoto() method.
Something important to note ... you use the node-raspistill library to take images. From my tests it takes **6.2** seconds on the average to take a picture. This is compared to the **1.1** seconds it takes with the previous spawn approach I used earlier. 6 seconds is large, is there any other reason we should go with node-raspistill?

File: README.md (hunk 2 above)
Comment (victordibia) [inline README.md]: Done.

File: .gitignore (hunk 1 above)
Comment (jweisz) [inline .gitignore]: tests belong in the `tests` folder. :D

i'm not a fan of blacklisting "special" directories only because they have code that you're trying to hide.

File: lib/tjbot.js (hunk 3 above)
Comment (jweisz) [inline lib/tjbot.js]: could this just be because we're not giving that "timeout" argument?

no reason in particular for using this node package, it just seemed easier than other methods (including invoking `raspistill` ourselves).  however, if it's the wrong choice, i'm not against going back to manual invocation (in a separate PR).

File: README.md [hunk 4]
```diff
@@ -390,7 +390,25 @@ tj.read().then(function(texts) {
...
-Sample resposne:
+Sample response:
+
+```
+TBD
+```
+
+### tj.takePhoto(filePath)
+
+Takes an argument of the path where the image file should be stored. If filepath is null, image is stored in a temporary location. This method returns the path where file was saved.
```
Comment (jweisz) [inline README.md]: should be `filePath`

Comment (jweisz) [inline README.md]: ... in "If `filePath` is null"

File: lib/tjbot.js [hunk 5]
```diff
@@ -197,6 +197,7 @@ TJBot.prototype._setupCamera = function() {
         height: this.configuration.see.camera.height,
         noPreview: true,
         encoding: 'jpg',
+        outputDir: './',
```
Comment (jweisz) [inline lib/tjbot.js]: `'./'` or `'/tmp'` ?

File: lib/tjbot.js [hunk 6]
```diff
@@ -735,31 +736,42 @@ TJBot.prototype.see = function() {
     var self = this;
 
     return new Promise(function(resolve, reject) {
-        temp.open('tjbot', function(err, info) {
-            if (err) {
-                reject("error: could not open temporary file for writing at path: " + info.path);
-            }
+        winston.verbose("TJBot taking a photo ");
```
Comment (jweisz) [inline lib/tjbot.js]: extra space at the end of the string

File: lib/tjbot.js [hunk 7]
```diff
@@ -735,31 +736,42 @@ TJBot.prototype.see = function() {
...
-                        var result = response.images[0].classifiers[0].classes;
-                        resolve(result);
-                    }
-                });
-            });
+    // capture 'this' context
+    var self = this;
+
+    return new Promise(function(resolve, reject) {
+        winston.debug("sending image to Watson Visual Recognition to identify concepts");
```
Comment (jweisz) [inline lib/tjbot.js]: i would delete "to identify concepts", it's not entirely accurate

File: lib/tjbot.js [hunk 8]
```diff
@@ -774,28 +786,78 @@ TJBot.prototype.read = function() {
...
+        self.takePhoto().then(function(filePath) {
+            resolve(self.describePhotoText(filePath))
+        });
+    });
+}
+
+/*
+Describe text in photo by sending it to the Watson Visual Recognition Services.
+ */
+TJBot.prototype.describePhotoText = function(filePath) {
```
Comment (jweisz) [inline lib/tjbot.js]: i'm not totally happy with the name `describePhotoText`.  It's not really about _describing_, it's more about _reading_ or _recognizing_.  The Watson API calls it `recognizeText`, so should this method be called `recognizeTextInPhoto()` ?  (and then, instead of `describePhoto()` above we should call it `recognizeObjectsInPhoto()`?)

File: lib/tjbot.js [hunk 9]
```diff
@@ -774,28 +786,78 @@ TJBot.prototype.read = function() {
...
+
+    this._camera.setOptions({
+        outputDir: path,
+        fileName: name
     });
+
+    return new Promise(function(resolve, reject) {
+        self._camera.takePhoto().then(function(photobuffer) {
+            resolve(path + "/" + name + "." + self._camera.getOptions().encoding);
+        }).catch((error) => {
```
Comment (jweisz) [inline lib/tjbot.js]: we should standardize on one way of doing callbacks:

**option A**
```
}).catch(function(error) {
   ...
```

**option B**
```
}).catch((error) => {
...
```

i think i used **option A** in all the previous code, i'd like this to be consistent

File: lib/tjbot.js [hunk 10]
```diff
@@ -735,31 +736,42 @@ TJBot.prototype.see = function() {
...
-            self._captureImage(info.path).then(function(buffer) {
-                winston.debug("sending image to Watson Visual Recognition");
-                var params = {
-                    images_file: fs.createReadStream(info.path + ".jpg"),
-                    threshold: self.configuration.see.confidenceThreshold.object
-                };
+/*
+Describe concepts in photo by sending it to the Watson Visual Recognition Services.
+ */
+TJBot.prototype.describePhoto = function(filePath) {
```
Comment (jweisz) [inline lib/tjbot.js]: see comment below about calling this method `recognizeObjectsInPhoto()`

File: lib/tjbot.js (hunk 3 above)
Comment (victordibia) [inline lib/tjbot.js]: Ok .... will experiment with that later.

File: lib/tjbot.js (hunk 5 above)
Comment (victordibia) [inline lib/tjbot.js]: Above was changed so that takePhoto("test.jpg") saves the file to the current directory.
When no argument is provided, it is saved to /tmp.

File: lib/tjbot.js (hunk 6 above)
Comment (victordibia) [inline lib/tjbot.js]: removed.

File: lib/tjbot.js (hunk 10 above)
Comment (victordibia) [inline lib/tjbot.js]: Updated

File: lib/tjbot.js (hunk 8 above)
Comment (victordibia) [inline lib/tjbot.js]: This has been updated.

File: lib/tjbot.js (hunk 9 above)
Comment (victordibia) [inline lib/tjbot.js]: Agreed. We have used option A normally, ... updatiing this.

File: README.md [hunk 11]
```diff
@@ -398,20 +398,20 @@ TBD
...
 ### tj.takePhoto(filePath)
 
-Takes an argument of the path where the image file should be stored. If filepath is null, image is stored in a temporary location. This method returns the path where file was saved.
+Takes an argument of the path where the image file should be saved. If `filePath` is null, image is stored in a temporary location. This method returns the location `targetPath` where image file was saved.
 
 Sample usage:
 
 ```
-tj.takePhoto(path).then(function(filePath) {
+tj.takePhoto(filePath).then(function(targetPath) {
```
Comment (jweisz) [inline README.md]: `targetPath` sounds like the target of where you want the photo to be saved, and `filePath` sounds like the place where the file was actually saved.  Does it make sense to flip these?

```
tj.takePhoto(targetPath).then(function(filePath) {
```

File: README.md [hunk 12]
```diff
@@ -398,20 +398,20 @@ TBD
...
     ...
 });
 ```
 
 Sample response:
 
 ```
-TBD
+test.jpg
```
Comment (jweisz) [inline README.md]:
```
targetPath = "test.jpg"
```

?

File: lib/tjbot.js [hunk 13]
```diff
@@ -736,24 +736,24 @@ TJBot.prototype.see = function() {
...
         self.takePhoto().then(function(filePath) {
-            resolve(self.describePhoto(filePath))
+            resolve(self.recognizeObjectsInPhoto(filePath))
         });
     });
 }
 
 /*
-Describe concepts in photo by sending it to the Watson Visual Recognition Services.
+Describe photo by sending it to the Watson Visual Recognition Services.
```
Comment (jweisz) [inline lib/tjbot.js]: "Service" should be singular.

File: lib/tjbot.js [hunk 14]
```diff
@@ -842,8 +842,9 @@ TJBot.prototype.takePhoto = function(filePath) {
     };
 
     winston.debug("capturing image at path: " + filePath);
-    path = filePath.substring(0, filePath.lastIndexOf("/"));
+    path = filePath.lastIndexOf("/") > 0 ? filePath.substring(0, filePath.lastIndexOf("/")) : ""; // save to current dir if no directory provided.
```
Comment (jweisz) [inline lib/tjbot.js]: For saving to the current directory, should that `""` be `"./"` instead?

Comment (victordibia) [inline lib/tjbot.js]: The lib intepretes "" as current directory.

File: README.md (hunk 12 above)
Comment (victordibia) [inline README.md]: updated.

File: README.md (hunk 11 above)
Comment (victordibia) [inline README.md]: makes sense. updated.
